{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "56ecc63f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.9/site-packages/IPython/core/magics/pylab.py:159: UserWarning: pylab import has clobbered these variables: ['record']\n",
      "`%matplotlib` prevents importing * from pylab and numpy\n",
      "  warn(\"pylab import has clobbered these variables: %s\"  % clobbered +\n"
     ]
    }
   ],
   "source": [
    "# Importing necessary packages, may need to download cyvcf2 and numpy joblib to enhance performance\n",
    "%pylab inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from cyvcf2 import VCF\n",
    "\n",
    "# Load phenotype data\n",
    "ptdata = pd.read_csv(os.path.expanduser(\"~/public/lab3/lab3_gwas.phen\"), delim_whitespace=True, names=[\"FID\",\"IID\",\"LDL\"])\n",
    "\n",
    "# Load genotype data\n",
    "vcf_file = os.path.expanduser(\"~/public/lab3/lab3_gwas.vcf.gz\")\n",
    "vcf_reader = VCF(vcf_file)\n",
    "\n",
    "# Extract sample IDs\n",
    "samples = vcf_reader.samples\n",
    "\n",
    "# Extract sample IDs & associated genotype data\n",
    "genotype_data = []\n",
    "variant_ids = []\n",
    "\n",
    "for record in vcf_reader:\n",
    "    variant_ids.append(record.ID)\n",
    "    genotypes = [gt[0] + gt[1] for gt in record.genotypes]\n",
    "    genotype_data.append(genotypes)\n",
    "\n",
    "# Convert genotype data to pandas DataFrame\n",
    "genotype_data = np.array(genotype_data).T\n",
    "genotype_df = pd.DataFrame(genotype_data, columns=variant_ids, index=samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f29da9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs11252127</th>\n",
       "      <th>rs7909677</th>\n",
       "      <th>rs11591988</th>\n",
       "      <th>rs12768206</th>\n",
       "      <th>rs10904561</th>\n",
       "      <th>rs7917054</th>\n",
       "      <th>rs7906287</th>\n",
       "      <th>rs9419557</th>\n",
       "      <th>rs9286070</th>\n",
       "      <th>rs11253562</th>\n",
       "      <th>...</th>\n",
       "      <th>rs2606358</th>\n",
       "      <th>rs2739260</th>\n",
       "      <th>rs2229949</th>\n",
       "      <th>rs3750508</th>\n",
       "      <th>rs3750510</th>\n",
       "      <th>rs9777369</th>\n",
       "      <th>rs11137376</th>\n",
       "      <th>rs17583562</th>\n",
       "      <th>rs11137379</th>\n",
       "      <th>rs9314655</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA06984</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA06989</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA12878</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA18489</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA18504</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 834975 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         rs11252127  rs7909677  rs11591988  rs12768206  rs10904561  rs7917054  \\\n",
       "NA06984           0          0           0           1           0          1   \n",
       "NA06989           1          1           0           1           1          1   \n",
       "NA12878           0          0           0           1           0          1   \n",
       "NA18489           0          0           0           1           0          1   \n",
       "NA18504           0          0           0           0           0          0   \n",
       "\n",
       "         rs7906287  rs9419557  rs9286070  rs11253562  ...  rs2606358  \\\n",
       "NA06984          1          0          0           1  ...          2   \n",
       "NA06989          2          0          0           0  ...          1   \n",
       "NA12878          1          0          0           1  ...          1   \n",
       "NA18489          1          0          1           0  ...          0   \n",
       "NA18504          0          0          0           0  ...          0   \n",
       "\n",
       "         rs2739260  rs2229949  rs3750508  rs3750510  rs9777369  rs11137376  \\\n",
       "NA06984          2          2          0          2          0           0   \n",
       "NA06989          1          2          1          2          0           0   \n",
       "NA12878          1          2          1          2          0           0   \n",
       "NA18489          0          1          0          2          1           1   \n",
       "NA18504          0          0          0          2          2           0   \n",
       "\n",
       "         rs17583562  rs11137379  rs9314655  \n",
       "NA06984           0           0          0  \n",
       "NA06989           1           1          0  \n",
       "NA12878           0           0          0  \n",
       "NA18489           0           0          1  \n",
       "NA18504           0           0          2  \n",
       "\n",
       "[5 rows x 834975 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter SNPs based on criteria used by the Wellcome Trust Case Control Consortium (WTCCC) when conducting GWAS\n",
    "# The criteria for retaining a SNP are: HWE P-value ≥ 5.7 × 10−7, MSP ≤5% if MAF ≥ 5%, MSP ≤ 1%\n",
    "def filter_snps(genotype_df, maf_threshold = 0.05, missingness_proportion = 0.05):\n",
    "    \n",
    "    # Calculate minor alllele frequency (MAF)\n",
    "    allele_freq = genotype_df.apply(lambda x: np.sum(x) / (2 * len(x)), axis = 0)\n",
    "    maf = np.minimum(allele_freq, 1 - allele_freq)\n",
    "    \n",
    "    # Filter SNPs based on MAF\n",
    "    genotype_df = genotype_df.loc[:, maf > maf_threshold]\n",
    "    \n",
    "    # Filter SNPs by missingness proportion (MSP)\n",
    "    missingness = genotype_df.isnull().mean()\n",
    "    genotype_df = genotype_df.loc[:, missingness < missingness_proportion]\n",
    "    \n",
    "    return genotype_df\n",
    "\n",
    "# Follow same guidelines usde by WTCCC to filter individuals based on missingness\n",
    "def filter_indivs(genotype_df, missingness_proportion = 0.05):\n",
    "    missingness = genotype_df.isnull().mean(axis = 1)\n",
    "    genotype_df = genotype_df.loc[missingness < missingness_proportion, :]\n",
    "    return genotype_df\n",
    "\n",
    "# Perform QC uby filtering SNPs, individuals, and based on HWE p-value\n",
    "genotype_df = filter_snps(genotype_df)\n",
    "genotype_df = filter_indivs(genotype_df)\n",
    "\n",
    "# Display the first few rows of the filtered genotype data\n",
    "genotype_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "56e19b91",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "exception calling callback for <Future at 0x7f0694f91340 state=finished raised BrokenProcessPool>\n",
      "joblib.externals.loky.process_executor._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/externals/loky/process_executor.py\", line 404, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/opt/conda/lib/python3.9/multiprocessing/queues.py\", line 108, in get\n",
      "    if not self._rlock.acquire(block, timeout):\n",
      "KeyboardInterrupt\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/externals/loky/_base.py\", line 625, in _invoke_callbacks\n",
      "    callback(self)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/parallel.py\", line 359, in __call__\n",
      "    self.parallel.dispatch_next()\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/parallel.py\", line 792, in dispatch_next\n",
      "    if not self.dispatch_one_batch(self._original_iterator):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/parallel.py\", line 859, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/parallel.py\", line 777, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 531, in apply_async\n",
      "    future = self._workers.submit(SafeFunction(func))\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/externals/loky/reusable_executor.py\", line 177, in submit\n",
      "    return super(_ReusablePoolExecutor, self).submit(\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/joblib/externals/loky/process_executor.py\", line 1102, in submit\n",
      "    raise self._flags.broken\n",
      "joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_177/2165523802.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m \u001b[0;31m# Apply HWE filtering\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 48\u001b[0;31m \u001b[0mfiltered_genotype_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfilter_snps_hwe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenotype_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhwe_threshold\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1e-6\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     49\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[0;31m# Display the shape of the filtered genotype DataFrame\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_177/2165523802.py\u001b[0m in \u001b[0;36mfilter_snps_hwe\u001b[0;34m(genotype_df, hwe_threshold, n_jobs)\u001b[0m\n\u001b[1;32m     37\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0msnp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp_value\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m     results = Parallel(n_jobs=n_jobs)(\n\u001b[0m\u001b[1;32m     40\u001b[0m         \u001b[0mdelayed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocess_snp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnp\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msnp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenotype_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m     )\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/joblib/parallel.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1052\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1053\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretrieval_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1054\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mretrieve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1055\u001b[0m             \u001b[0;31m# Make sure that we get a last message telling us we are done\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1056\u001b[0m             \u001b[0melapsed_time\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_start_time\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/joblib/parallel.py\u001b[0m in \u001b[0;36mretrieve\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    931\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    932\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'supports_timeout'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 933\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    934\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    935\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_output\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjob\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/joblib/_parallel_backends.py\u001b[0m in \u001b[0;36mwrap_future_result\u001b[0;34m(future, timeout)\u001b[0m\n\u001b[1;32m    540\u001b[0m         AsyncResults.get from multiprocessing.\"\"\"\n\u001b[1;32m    541\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 542\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfuture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    543\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mCfTimeoutError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    544\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mTimeoutError\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/concurrent/futures/_base.py\u001b[0m in \u001b[0;36mresult\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    438\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__get_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    439\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 440\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_condition\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    441\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    442\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mCANCELLED\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mCANCELLED_AND_NOTIFIED\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    310\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m    \u001b[0;31m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    311\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 312\u001b[0;31m                 \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    313\u001b[0m                 \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    314\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from joblib import Parallel, delayed\n",
    "from scipy.stats import chi2_contingency\n",
    "# Function to calculate HWE p-value\n",
    "def hwe_p_value(genotype_counts):\n",
    "    obs_homo1 = genotype_counts.get(0, 0)\n",
    "    obs_het = genotype_counts.get(1, 0)\n",
    "    obs_homo2 = genotype_counts.get(2, 0)\n",
    "    n = obs_homo1 + obs_het + obs_homo2\n",
    "\n",
    "    if n == 0:\n",
    "        return 1.0  # Avoid division by zero\n",
    "\n",
    "    p = (2 * obs_homo1 + obs_het) / (2 * n)\n",
    "    q = 1 - p\n",
    "\n",
    "    exp_homo1 = n * p * p\n",
    "    exp_het = 2 * n * p * q\n",
    "    exp_homo2 = n * q * q\n",
    "\n",
    "    # Skip SNPs where the expected count for any genotype is less than 1\n",
    "    if min(exp_homo1, exp_het, exp_homo2) < 1:\n",
    "        return 1.0\n",
    "\n",
    "    observed = [obs_homo1, obs_het, obs_homo2]\n",
    "    expected = [exp_homo1, exp_het, exp_homo2]\n",
    "\n",
    "    chi2, p_value = chi2_contingency([observed, expected])[:2]\n",
    "\n",
    "    return p_value\n",
    "\n",
    "# Function to filter SNPs based on HWE\n",
    "def filter_snps_hwe(genotype_df, hwe_threshold=1e-6, n_jobs=-1):\n",
    "    def process_snp(snp):\n",
    "        genotypes = genotype_df[snp].dropna()\n",
    "        genotype_counts = genotypes.value_counts().to_dict()\n",
    "        p_value = hwe_p_value(genotype_counts)\n",
    "        return (snp, p_value)\n",
    "\n",
    "    results = Parallel(n_jobs=n_jobs)(\n",
    "        delayed(process_snp)(snp) for snp in genotype_df.columns\n",
    "    )\n",
    "\n",
    "    snps_to_keep = [snp for snp, p_value in results if p_value >= hwe_threshold]\n",
    "\n",
    "    return genotype_df[snps_to_keep]\n",
    "\n",
    "# Apply HWE filtering\n",
    "filtered_genotype_df = filter_snps_hwe(genotype_df, hwe_threshold=1e-6)\n",
    "\n",
    "# Display the shape of the filtered genotype DataFrameg\n",
    "print(f\"Shape of genotype data before HWE filtering: {genotype_df.shape}\")\n",
    "print(f\"Shape of genotype data after HWE filtering: {filtered_genotype_df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37ee1906",
   "metadata": {},
   "outputs": [],
   "source": [
    "# THE ONE THAT WORKED\n",
    "from scipy.stats import t\n",
    "from sklearn.linear_model import LinearRegression\n",
    "# Function for association testing using linear regression\n",
    "def linear_regression_association_test(genotype_df, phenotype_data):\n",
    "    results = []\n",
    "\n",
    "    # Iterate over each SNP in the filtered genotype DataFrame\n",
    "    for snp in genotype_df.columns:\n",
    "        # Extract genotype values for the current SNP\n",
    "        genotype_values = genotype_df[snp].values.reshape(-1, 1)\n",
    "\n",
    "        # Extract phenotype values corresponding to the genotypes\n",
    "        phenotype_values = phenotype_data[\"LDL\"].values.reshape(-1, 1)\n",
    "\n",
    "        # Fit linear regression model\n",
    "        model = LinearRegression().fit(genotype_values, phenotype_values)\n",
    "\n",
    "        # Get the coefficient and standard error\n",
    "        coef = model.coef_[0][0]\n",
    "        std_err = np.sqrt(np.mean((phenotype_values - model.predict(genotype_values))**2))  # Standard error of the coefficient\n",
    "\n",
    "        # Calculate the t-statistic and its p-value\n",
    "        t_statistic = coef / std_err\n",
    "        dof = len(phenotype_values) - 2  # Degrees of freedom\n",
    "        p_value = 2 * (1 - t.cdf(abs(t_statistic), df=dof))\n",
    "\n",
    "        # Append SNP and p-value to results\n",
    "        results.append((snp, p_value))\n",
    "\n",
    "    # Create DataFrame from results\n",
    "    results_df = pd.DataFrame(results, columns=['SNP', 'p-value'])\n",
    "    \n",
    "    return results_df\n",
    "\n",
    "results = linear_regression_association_test(genotype_data, phenotype_data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
